src/Avalonia.Controls/Primitives/Popup.cs

```diff
@@ -472,6 +503,7 @@ void DeferCleanup(IDisposable? disposable)
             WindowManagerAddShadowHintChanged(popupHost, WindowManagerAddShadowHint);
 
             popupHost.Show();
+            IsConstrainedToRootBounds = popupHost is OverlayPopupHost;
```

Comment:
I don't think this is correct. IsConstrainedToRootBounds (by my limited understanding) indicates when the flyout is actually constrained. This would be, for example, the flyout/popup requests position below an element but is instead shown above it due to the bounds constraints. You are just always returning true depending on how the popup host is implemented?

I would expect to see some bounds calculations some place with this property.

Reply:
That's why I found this naming confusing.
This property seems to do nothing with positioning and constraining. On this gif I used two Button.Flyouts with Placement.Bottom and different values of ShouldConstrainToRootBounds. As you can see, value of IsConstrainedToRootBounds is the same for requested placement (Bottom) and for overflowing placement (Right and Top on gif).
https://gist.github.com/maxkatz6/786403357baae5d31eb2d1448f2ea59f
![123](https://user-images.githubusercontent.com/3163374/110230890-1ed64680-7ee2-11eb-868e-34a5a75bba52.gif)

Reply:
From my observation, ```IsConstrainedToRootBounds``` is False only when ```ShouldConstrainToRootBounds``` was False and device actually supports popups outside of parent window (so it will be always True on Mobile for instance). In all other cases this property has "True" value.

Reply:
I have to think this is either a bug in WinUI or a result of a lazy initial implementation. I'm not sure WinUI functionality that doesn't make sense should be replicated. For Avalonia it might be better to just have one `IsConstrainedToRootBounds` { get; set; } and be done with it.

Reply:
As for me it's pretty clear, that the difference between "true" and "false" is only in fact is popup using separated window or is it rendered in the overlay layer. And user can only set "hint" value, that can be ignored by the platform.
MS docs and random [comments](https://github.com/microsoft/microsoft-ui-xaml/issues/1074#issuecomment-514392995) from developers tells that.

Reply:
I understand that distinction. I had two points:

 1. In your graphic the value of ShouldConstrainToRootBounds is always equal to IsConstrainedToRootBounds. Maybe Windows can't always honor the request but in Avalonia I was thinking you will always be able to honor it as you decide how to draw the popup. Then why have two properties always equal? At the time I wasn't thinking of the case where the user requests ShouldConstrainToRootBounds=False on mobile devices though.
 2. IsConstrainedToRootBounds returns True even if the overlay isn't physically constrained to the root bounds. This kind of makes the property useless doesn't it? I would expect, as I said earlier, for it to represent how the overlay is actually laid out and if the bounds have to be artificially adjusted to fit within a window. This information could be useful for adjusting the view in that case (maybe hide certain elements for example).

Point 2 is still valid and point 1 was partly my misunderstanding at first. However, it is still entirely possible to combine these two properties together into 1. Set will set a request and get will get the actual state.

Can you mark as unresolved until we get a 3rd opinion?

Reply:
Microsoft docs also seem to have been modified to make UWP behavior 'by design' when it contradicts the sentence above.

> To check if the popup is actually constrained to the root bounds, check the IsConstrainedToRootBounds property.
>
>In a UWP app on desktop, when this property is true, the popup is shown within the main XAML window handle (HWND). When this property is false, the popup is shown in its own top level HWND. In this case, the popup might be positioned to extend beyond the main app window.

Reply:
Last comment for today, I promise :) The word 'Constrained' might be contributing to the confusion here. The definition of the word itself implies that size or location is forcefully adjusted when true. And we know from your tests that's not how it works. It may be better to use 'Within'. 'IsWithinRootBounds' along with my logic adjustments above would makes MUCH more sense in my mind.

Reply:
> in Avalonia I was thinking you will always be able to honor it as you decide how to draw the popup

It's not true. If developer asked for ```ShouldConstrainToRootBounds=false```, but overlay popups was enabled globally (Win32PlatformOptions.OverlayPopups=true) or application is running on platform, where window-based popups are not supported (mobile), then developers request will be ignored and ```IsConstrainedToRootBounds``` will be ```true``` instead of ```false```.

Reply:
> IsConstrainedToRootBounds returns True even if the overlay isn't physically constrained to the root bounds. This kind of makes the property useless doesn't it.

From my understanding of "constrained" definition, it's not mandatory for popup to be fit inside of parent window without any cropping, but more like positioning rule. And UWP returns "True" when application can follow this rule from platform restrictions perspective, ignoring actual positioning status.
Although I totally agree with you about confusion. In my first comment I already said, that I don't like this naming.